### DISCUSSION TITLE
nixos-with-flakes/update-the-system

### DISCUSSION THREAD

**Discussion by giscus** [ryan4yin/nixos-and-flakes-book#154] [Announcements]
# nixos-with-flakes/update-the-system

An unofficial and opinionated book for beginners

https://nixos-and-flakes.thiscute.world/nixos-with-flakes/update-the-system

**Comment by B33th0v3n960** [comment]
Is there a way to automate update every month or should this be done manually?

**Reply by athian** [reply]
Yes there is, with system.autoUpgrade: https://search.nixos.org/options?channel=unstable&query=system.autoUpgrade
Official wiki has an example of what it would look like: https://wiki.nixos.org/wiki/Automatic_system_upgrades#Configuration

**Comment by exiquio** [comment]
Is it possible to list or view what will be or has been update package-wise?

**Reply by athian** [reply]
You can compare your current running system and what appears after `nixos-rebuild build` with `nix store diff-closures /run/current-system ./result` command. Or between your current and some of the previous generations that you can find in `/nix/var/nix/profiles`.
You can also do it automatically after switch: I have this snippet in my configuration:
```
system.activationScripts.diff = ''
  if [[ -e /run/current-system ]]; then
    ${pkgs.nix}/bin/nix store diff-closures /run/current-system "$systemConfig"
  fi
'';
```
and there is also this tool, nvd, that does a little more: https://khumba.net/projects/nvd/

**Comment by stegrams** [comment]
I'm a little confused about the permissions here.
When I initialized flakes on my ```/etc/nixos``` directory, I had to use ```sudo``` or else I had an error opening file permissions denied.
Have I done something wrong?
Because now I also have permission error when I try to update without ```sudo``` as instructed.
Also, why do I have to add ```--flake .``` when the rebuild command will prioritize reading the flakes.nix, as noted in chapter [Enabling NixOS with Flakes](https://nixos-and-flakes.thiscute.world/nixos-with-flakes/nixos-with-flakes-enabled#switch-to-flake-nix)
Thank you.

**Reply by ryan4yin** [reply]
1. The `/etc/nixos` directory is owned by and only writable by `root`. Therefore, if your flake is located in this directory, you'll need to use `sudo` to modify any configuration files.

2. The `--flake .` flag is not required if your configuration is located at `/etc/nixos`. I've updated this section to clarify that.
